crates/paimon/src/fileindex/file_index_format.rs

```diff
@@ -0,0 +1,486 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{collections::HashMap, io::SeekFrom};
+
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter};
+
+pub const MAGIC: u64 = 1493475289347502;
+pub const EMPTY_INDEX_FLAG: i32 = -1;
+
+/// File index file format. Put all columns and offsets in the header.
+///
+/// ```
+///   _____________________________________    _____________________
+/// ｜     magic    ｜version｜head length  ｜
+/// ｜-------------------------------------｜
+/// ｜            column number            ｜
+/// ｜-------------------------------------｜
+/// ｜   column 1        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜            HEAD
+/// ｜   column 2        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜  redundant length ｜redundant bytes  ｜
+/// ｜-------------------------------------｜    ---------------------
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜             BODY
+/// ｜                BODY                 ｜
+/// ｜_____________________________________｜    _____________________
+/// ```
+///
+/// - `magic`: 8 bytes long
+/// - `version`: 4 bytes int
+/// - `head length`: 4 bytes int
+/// - `column number`: 4 bytes int
+/// - `column x`: var bytes utf (length + bytes)
+/// - `index number`: 4 bytes int (how many column items below)
+/// - `index name x`: var bytes utf
+/// - `start pos`: 4 bytes int
+/// - `length`: 4 bytes int
+/// - `redundant length`: 4 bytes int (for compatibility with later versions, in this version, content is zero)
+/// - `redundant bytes`: var bytes (for compatibility with later version, in this version, is empty)
+/// - `BODY`: column index bytes + column index bytes + column index bytes + .......
+
+#[derive(Debug)]
+pub struct FileIndexFormat;
```

Comment:
Using an empty struct to provide `create_writer`/`create_reader` is not a common pattern. I suggest implementing `FileIndexReader` and `FileIndexWriter` that can accept a `FileIO` instead.

Reply:
Maybe what is needed here is called FileIndexFormatReader and FileIndexFormatWriter.